### PR TITLE
C++ visibility support

### PR DESCRIPTION
**Pull request by j-rivero**
This PR implements the [control of binary symbols visibility](https://gcc.gnu.org/wiki/Visibility) in FCL.

It adds a new `FCL_VISIBLE` preprocessor define to any class, struct or function that needs to be exposed publicly to third party software. Since the number of files in this library is not small, I've patched all what I found needed to compile the whole test suite. It would be good to use another third party software using FCL extensively to test that I did not forget about any of them.

The PR also adds the `FCL_HIDE_ALL_SYMBOLS` option (OFF by the default). This option covers a use case we found in Drake where a software uses FCL (linked statically) but does not expose it in the public API (headers). In this scenario the best option to avoid any possible conflict with other versions of fcl is to hide all the symbols from the public ABI.

Numbers before and after:

| Version | Number of symbols | libfcl size |
| ----------|--------------------------|-------------|
| current | 18942                      | 11922 Kb     |
|with visibility from this PR | 16530 | 11257Kb |
|no symbols | 157 |  9012Kb |

Summary: **13% symbols less** and **6% size less** . 

The PR will alter the ABI so a new bump in the FCL version prior to release it would be ideal.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -12,6 +12,27 @@ set_target_properties(${PROJECT_NAME} PROPERTIES
   VERSION ${FCL_VERSION}
   SOVERSION ${FCL_ABI_VERSION})
 
+# Hide symbols by default
+#################################################
+# Macro to check for visibility capability in compiler
+# Original idea from: https://gitorious.org/ferric-cmake-stuff/
+macro (check_gcc_visibility)
```
Comment (mamoll) [inline src/CMakeLists.txt]: This also works for clang (and maybe other compilers), so maybe rename to check_compiler_visibility?

File: include/fcl/common/visibility.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/** \def FCL_VISIBLE
+ * Use to represent "symbol visible" if supported
+ */
+
+/** \def FCL_HIDDEN
+ * Use to represent "symbol hidden" if supported
+ */
+
+#if defined _WIN32 || defined __CYGWIN__
+  #ifdef BUILDING_DLL
```
Comment (traversaro) [inline include/fcl/common/visibility.h]: In Gazebo, the usage of the same `BUILDING_DLL` macro for different libraries created several problems (see https://bitbucket.org/osrf/gazebo/issues/2262/symbol-visibility-issues-in-windows). I would recommend to use a fcl-specific macro and set the [`DEFINE_SYMBOL`](https://cmake.org/cmake/help/v3.5/prop_tgt/DEFINE_SYMBOL.html) property of the library appropriately.

Comment (traversaro) [reply include/fcl/common/visibility.h]: The header in this form is breaking the static builds in MSVC, that is the default behaviour when building the library in MSVC.

File: include/fcl/narrowphase/detail/convexity_based_algorithm/epa.h [hunk 3]
```diff
@@ -52,9 +52,9 @@ namespace detail
 //static const size_t EPA_MAX_ITERATIONS = 255;
 // TODO(JS): remove?
 
-/// @brief class for EPA algorithm
+/// @brief class FCL_VISIBLE for EPA algorithm
```
Comment (traversaro) [inline include/fcl/narrowphase/detail/convexity_based_algorithm/epa.h]: There are a few spurious insertion of the `FCL_VISIBLE` macro in the documentation, they can be easily found by searching for `FCL_VISIBLE for`.

File: include/fcl/common/visibility.h (hunk 2 above)
Comment (j-rivero) [reply include/fcl/common/visibility.h]: Oh sorry forgot about this issue, thanks for the reminder Silvio. Could please give a look at this patch https://github.com/j-rivero/fcl/commit/e509a3d1659591bb4bacc98638ff74e505c5fee7 and let me know if it is correct? Unfortunately I don't have a Windows/MSVC15 setup at hand to test.

**Comment by j-rivero** [comment]
Ouk, sorry I've pushed the mess of my working copy to the repo (reverts on reverts). I'll open a new pull request with a clean history and close this one.

**Comment by j-rivero** [comment]
Excuse me for the change of PR. Let's follow in #233

File: include/fcl/common/visibility.h (hunk 2 above)
Comment (traversaro) [reply include/fcl/common/visibility.h]: Commented in https://github.com/j-rivero/fcl/commit/e509a3d1659591bb4bacc98638ff74e505c5fee7 .

Comment (traversaro) [reply include/fcl/common/visibility.h]: Sorry, I actually commented in https://github.com/flexible-collision-library/fcl/commit/e509a3d1659591bb4bacc98638ff74e505c5fee7 .

**Pull request closed without merging**